Improve lima 2D controller + add 1D controller

tiagocoutinho opened the pull request:
This PR proposes:
* a sardana 1D controller for a lima detector
* working versions of both 1D and 2D controller with `ValueRefEnabled = False`

Commit:
Working version with Lima 1D and value ref disabled working

rhomspuron commented:
Nice :)

One question, Does this PR solve the issue #5? If not, maybe it is a good moment to do it.

jordiandreu commented on sardana_limaccd/ctrl/LimaCCDCtrl.py:
```diff
@@ -58,121 +63,144 @@
+    DTypeSize = (1, 2, 4, None, 1, 2, 4)
+
+    def __init__(self):
+        super().__init__('<IHHIIHHHHHHHHIIIIIIII')
```
Should this controllers be python2 compatible? In this case, we need to fix the super calls, adding the class and self instance as arguments.

rhomspuron replied:
Yes, the idea is to have the same version on py2 and py3 if it does not block the development.

tiagocoutinho replied:
Ok I can do that.
I completely forgot this was to be compatible with py2

tiagocoutinho replied:
Ok, latest commit should fix plenty of things including this one

jordiandreu commented:
I'm currently also testing the code in sardana-py3 (also in python2) and there is an exception that need to be addressed:
`  File "/home/jandreu/Development/ALBA-CS/sardana/src/sardana/pool/poolbasechannel.py", line 446, in extend_value_ref_buffer
    if len(value_refs) == 0:
TypeError: object of type 'SardanaValue' has no len()`

tiagocoutinho commented:
@rhomspuron :
Thanks, I plan to use it within ALBA MSPD to control the mythen detector.
Actually, this fixes #5 in the new supported feature of `ValueRefEnabled=False`.
I propose that to fix #5 we create another PR. I have some idea to fix it but it needs to be tested with old a new versions of Lima

tiagocoutinho commented:
@jordiandreu ,
Thanks for testing!

> I'm currently also testing the code in sardana-py3 (also in python2) and there is an exception that need to be addressed:
File "/home/jandreu/Development/ALBA-CS/sardana/src/sardana/pool/poolbasechannel.py", line 446, in extend_value_ref_buffer if len(value_refs) == 0: TypeError: object of type 'SardanaValue' has no len()

I suspect a bug in sardana...

tiagocoutinho commented:
@jordiandreu,
The error you got was for Lima2D controller or for a Lima1D controller? Were you testing with `ValueRefEnabled=False`?

jordiandreu commented:
> @jordiandreu,
> The error you got was for Lima2D controller or for a Lima1D controller? Were you testing with `ValueRefEnabled=False`?

@tiagocoutinho ,
I'm using Lima2D controller with `ValueRefEnabled=True`
The measurement group has a single channel, which corresponds to the first axis of the Lima2D controller.

reszelaz commented:
> I'm using Lima2D controller with ValueRefEnabled=True
The measurement group has a single channel, which corresponds to the first axis of the Lima2D controller.

In principle, this should work, at least for the dummy 2D controller it worked not so long time ago, see example from https://github.com/sardana-org/sardana/pull/1510#issuecomment-786265662.

tiagocoutinho commented:
@jordiandreu ,
which type of synchronization do you have?
I think I can reproduce your error in SoftwareStart

jordiandreu commented:
> @jordiandreu ,
> which type of synchronization do you have?
> I think I can reproduce your error in SoftwareStart

Indeed, I was using Software synchronization and Start trigger. Moreover, a similar problem arises by using a dummy trigger gate controller as synchronizer and Start trigger...

rhomspuron commented:
> Indeed, I was using Software synchronization and Start trigger. Moreover, a similar problem arises by using a dummy trigger gate controller as synchronizer and Start trigger...

If you use a dummy trigger, it will fail because Lima will wait for HW trigger.

tiagocoutinho committed:
Add logging to PrepareOne, LoadOne

tiagocoutinho committed:
Fix ReadOne

tiagocoutinho committed:
First version

tiagocoutinho commented:
As discussed with @rhomspuron and @jordiandreu I change this PR to include all improvements including adding a 1D controller.

## Improvements

* [x] Add MasterTrigger property (default=False): allows to set the detector as master trigger
* [x] Check detector capabilites
* [x] Use internal trigger multi when detector supports it or internal trigger otherwise
* [x] Fix latency calculation based on synchro mode
* [x] Fix moving state on startup
* [x] In prepare, stop last acquisition if it did not finish well (ex: lost of a trigger)
* [x] Fix MaxDimSize
* [x] Only allow 2D to work in `ValueRefEnabled=True`

## 1D Controller

* [x] Support ALBA use cases: mythen (dectris) and Xspress3 (quantum detectors)
* [x] Support multiple axes. Each axis number corresponds to 2D row index
* [x] Only allow 1D to work in `ValueRefEnabled=False` (at least until [sardana #1445](https://github.com/sardana-org/sardana/issues/1445) is fixed

tiagocoutinho commented:
> In principle, this should work, at least for the dummy 2D controller it worked not so long time ago, see example from sardana-org/sardana#1510 (comment).

I confirm that it was a problem with my implementation.

tiagocoutinho committed:
Implement improvements from #11

## Requirements

* [x] Maintain API compatibility with previous version
* [x] Maintain support for python 2

## Improvements

* [x] Add MasterTrigger property (default=False): allows to set the detector as master trigger
* [x] Check detector capabilites
* [x] Use internal trigger multi when detector supports it or internal trigger otherwise
* [x] Fix latency calculation based on synchro mode
* [x] Fix moving state on startup
* [x] In prepare, stop last acquisition if it did not finish well (ex: lost of a trigger)
* [x] Fix MaxDimSize
* [x] Only allow 2D to work in `ValueRefEnabled=True`

## 1D Controller

* [x] Support ALBA use cases: mythen (dectris) and Xspress3 (quantum detectors)
* [x] Support multiple axes. Each axis number corresponds to 2D row index
* [x] Only allow 1D to work in `ValueRefEnabled=False` (at least until [sardana #1445](https://github.com/sardana-org/sardana/issues/1445) is fixed

tiagocoutinho commented:
I have tested with a simulated mythen and the official Lima simulator in software start and software trigger. 
It seems to work.

I have pending to test:
* mythen on the beamline with hardware trigger. 
* environment with saving to NFS to check saving still works well in "slow" file systems

tiagocoutinho commented:
Currently, the latest `lima-tango-server` conda package (1.9.2) has some bugs using python 3.
I recommend using directly the latest from [lima tango server repo](https://github.com/esrf-bliss/Lima-tango-python) for testing.

Namely:
* readImageSeq does not work in python 3
* if detector has capabilities like shutter, the pytango 9 server fails to start

jordiandreu commented on sardana_limaccd/client.py:
```diff
@@ -0,0 +1,323 @@
+    DTypeSize = (1, 2, 4, None, 1, 2, 4)
+
+    def __init__(self):
+        super().__init__("<IHHIIHHHHHHHHIIIIIIII")
```
This has to be:
`super(LimaImageFormat, self).__init__("<IHHIIHHHHHHHHIIIIIIII")`

tiagocoutinho replied:
Agree. nice catch

jordiandreu commented on sardana_limaccd/client.py:
```diff
@@ -0,0 +1,323 @@
+        names, values = zip(*config.items())
+        self.lima[names] = values
+        if self.first_image_nb != 0:
+            self.lima["saving_index_next_number"] = -1
```
The correct key is saving_next_number
`self.lima["saving_next_number"] = -1`

tiagocoutinho replied:
Agree. Good one

jordiandreu commented on sardana_limaccd/client.py:
```diff
@@ -0,0 +1,323 @@
+            self.lima["saving_prefix"] = config["saving_prefix"]
+            t0 = time.monotonic()
+            saving_next_number = -1
+            # After settinge the prefix with saving mode = ABORT,
```
just a typo here

tiagocoutinho replied:
Ok, should be fixed

jordiandreu reviewed (changes requested):
By now, tested in sardana develop with python3.7. More test with python 2.7 ongoing.

Commit:
Fix latency time property

Commit:
Fix missing start in external_* modes

Commit:
Fix missing python 2 compatibility call to super

Commit:
Fix saving_next_number attribute name

Commit:
Fix typo comment

jordiandreu commented on sardana_limaccd/client.py:
```diff
@@ -0,0 +1,323 @@
+            self.lima["saving_next_number"] = -1
+            time.sleep(0.05) # empirical value?
+            self.lima["saving_prefix"] = config["saving_prefix"]
+            t0 = time.monotonic()
```
`time.monotonic` is not available for python < 3.3. Some third party modules implement this method for old python version (https://pypi.org/project/monotonic/) but maybe is better to do not add an extra dependency if can be easily avoided.

tiagocoutinho replied:
Yes, you are right.

I guess we can do something simple like:
```python
 monotonic = getattr(time, "monotonic", time.time)
```

jordiandreu replied:
Indeed, once ready I can test the code again.

jordiandreu commented on sardana_limaccd/client.py:
```diff
@@ -0,0 +1,323 @@
+            # Allow to set the First Image Number to any value different to
+            # 0, default value on LimaCCDs after writing the prefix with
+            # saving mode in Abort
+            while saving_next_number == -1 and time.monotonic() - t0 < 2.5:
```
Also here...

jordiandreu replied:
Done!

jordiandreu committed:
Add compatibility for python < 3.3

Monotonic is not available for old python versions.

Commit:
Simplify access to saving

Commit:
Add several trigger mode pre-calculation

Commit:
Remove repeated unused code

Commit:
Protect against multiple start in external trigger mode (ex: step scan)

Commit:
Make acquisition aware of nb of starts

Commit:
Handle ext_trig & multi start (ex: step scan)

In ext_trig, when multiple starts (ex: ascan) report a "temporary"
ready state when it is ready to report new frames

Commit:
Handle start in Acquisition

tiagocoutinho commented:
I had to add a protection for the case of hardware trigger (start/stop/gate) where there are multiple starts (ex: in a step scan)

I am sorry but I have to ask for another review.

tiagocoutinho commented:
Here are the tests I did today in the lab using a mythen 2 as a lima 1D controller:

ValueRefEnabled = FALSE | (which means lima saving = False; all frames stored by sardana)

source | type | Lima eq. | ct | ascan | timescan | ascanc
-- | -- | -- | -- | -- | -- | --
hw (tg1) | trigger | ext_trig_mult | OK | OK | OK | OK
hw (tg1) | start | ext_trig | OK | --- | OK | OK
hw (tg1) | gate | ext_gate | TODO (1)
soft | trigger | int_trig_mult | OK | OK | OK | OK
soft | start | int_trig | OK | --- | OK | OK
soft | gate | int_trig_mult | OK | OK | OK | OK

(1) I need to wait for a proper cable

jordiandreu reviewed (approved):
Nice work! I have validated this version using python2/python3 Sardana versions with a Lima Simulator 2D controller, with value_ref_enable=True and macros **ascanct** and **ct**.

tiagocoutinho commented:
Thanks for yet another review @jordiandreu.
On my side I am satisfied with the results. I don't expect to make any more changes.
I guess we could consider merging it now

rhomspuron reviewed (commented):
Nice work!!

Only one question:
The LimaCtrl.py has the same code than LimaCCDCtrl with different classes names, why we would like to have this?

tiagocoutinho committed:
Remove garbage

tiagocoutinho commented:
Nice catch @rhomspuron. LimaCtrl.py is just garbage that got into git by accident.
I just removed it

tiagocoutinho commented:
Ok, if nobody disagrees I can merge it then.
